### ISSUE TITLE
Localstack pro cannot execute lambda functions

### ISSUE THREAD

**Issue by securisec** [localstack/localstack#3782]
# Type of request: This is a ...

[x] bug report
[ ] feature request

# Detailed description
I saw in one of the issues that localstack pro supports custom authorizers for api gateway, so I signed up for a trial license to test out a custom authorizer, but when setting a valid `LOCALSTACK_API_KEY`, the lambda function, or any functions does not run and throws an error. This does not happen when using localstack community with the same exact setup. 

...

## Expected behavior
The lambda function to execute in pro just as in community.
...

## Actual behavior
The function executes correctly in community version, but the executor fails in pro
...

# Steps to reproduce
Image hash:
`sha256:a583eadb6ad2a5e30dfaeb16ad2d4b818b3a05fe445aceaf66f8d79519546b75`

## Command used to start LocalStack
docker-compose file
```
version: '3'

services:
  localstack:
    image: localstack/localstack
    container_name: localstack
    ports:
      - "4566:4566"
    environment:
      - SERVICES=ec2,serverless,kinesis,cloudfront,sqs,sns,ses,ssm,iam,acm,cognito
      - LAMBDA_EXECUTOR=docker
      - LAMBDA_REMOVE_CONTAINERS=true
      - DOCKER_HOST=unix:///var/run/docker.sock
      - LAMBDA_REMOTE_DOCKER=false
      # - LOCALSTACK_API_KEY=myapikey
      - HOST_TMP_FOLDER=/tmp/
      - DATA_DIR=/tmp/localstack/data
      - AWS_ACCESS_KEY_ID=test
      - AWS_SECRET_ACCESS_KEY=test
      - DEBUG=true
    volumes:
      - "./data:/tmp/localstack"
      # - "/tmp/localstack:/tmp/localstack"
      - "/var/run/docker.sock:/var/run/docker.sock"
      # - "/tmp/localstack_docker:/opt/code/localstack"
```

If the `LOCALSTACK_API_KEY` is commented out, everything works fine, but if I enable the comment, it will fail. In order to do a sanity check to make sure I am getting the Pro features, when I run `awslocal cognito-idp list-user-pools --max-results 1` when the api key enabled, it will return an empty array, but when i comment out the pro key, i get a 400 error. 
...

When using localstack community, the response is simply `sent` as expected from the function code. 

But when using a valid Pro apikey, the response/localstack logs are:

```
{
  "errorType": "InvocationException",
  "errorMessage": "Lambda process returned error status code: 127. Result: . Output:\nWARNING: Localhost DNS setting (--dns=127.0.0.1) may fail in containers.\ndocker: Error response from daemon: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: \"/tmp/localstack/b789d1b8.sh\": stat /tmp/localstack/b789d1b8.sh: no such file or directory: unknown.",
  "stackTrace": [
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_api.py\", line 552, in run_lambda\n    result = LAMBDA_EXECUTOR.execute(func_arn, func_details, event, context=context,\n",
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 174, in execute\n    return do_execute()\n",
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 166, in do_execute\n    return _run(func_arn=func_arn)\n",
    "  File \"/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py\", line 149, in wrapped\n    raise e\n",
    "  File \"/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py\", line 145, in wrapped\n    result = func(*args, **kwargs)\n",
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 157, in _run\n    raise e\n",
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 145, in _run\n    result = self._execute(func_arn, func_details, event, context, version)\n",
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 323, in _execute\n    result = self.run_lambda_executor(cmd, stdin, env_vars=environment, func_details=func_details)\n",
    "  File \"/opt/code/localstack/.venv/lib/python3.8/site-packages/localstack_ext/services/awslambda/lambda_extended.py\", line 261, in run_lambda_executor\n",
    "  File \"/opt/code/localstack/localstack/services/awslambda/lambda_executors.py\", line 229, in run_lambda_executor\n    raise InvocationException('Lambda process returned error status code: %s. Result: %s. Output:\\n%s' %\n"
  ]
}
``` 

## Client code (AWS SDK code snippet, or sequence of "awslocal" commands)
Lambda function:
```go
package main

import (
	"context"

	"github.com/aws/aws-lambda-go/events"
	"github.com/aws/aws-lambda-go/lambda"
)

func handler(ctx context.Context, event events.APIGatewayProxyRequest) (events.APIGatewayProxyResponse, error) {
	return events.APIGatewayProxyResponse{
		Body:       "sent",
		StatusCode: 200,
	}, nil
}

func main() {
	lambda.Start(handler)
}
```
...

## Feature request:
Please consider making lambda custom authorizers a part of localstack community.

**Comment by yangzhang21** [comment]
any update on this? I see an issue similar......
 ERROR: 'docker run --user=root --cap-add=NET_ADMIN --entrypoint=/tmp/301c00b4.sh --dns 127.0.0.1 -e AWS_LAMBDA_FUNCTION_MEMORY_SIZE=123 -e _LAMBDA_RUNTIME=python3.8 -e .............................exit code 127; output: b'WARNING: Localhost DNS setting (--dns=127.0.0.1) may fail in containers.\n85db1691b216cd3748177b9562492caf3f1be63ec3bfca7220d32fd35a820806\ndocker: Error response from daemon: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "/tmp/301c00b4.sh": stat /tmp/301c00b4.sh: no such file or directory: unknown.\n'

**Comment by whummer** [comment]
Thanks for reporting @securisec @yangzhang21 . We've recently pushed a couple of enhancements for Lambda execution - can you please pull the `latest` Docker image and give it another try? 

Please ensure that the `HOST_TMP_FOLDER` environment variable is defined properly - it should point to the absolute path of the tmp folder on your host system (outside Docker) that gets mounted as `/tmp/localstack` into the container. In your case, looks like this should be the absolute path to the `./data` folder.

Also, you can use this command to validate your local configurations defined in `docker-compose.yml`:
```
$ pip install --upgrade localstack
$ localstack config validate
```
Please keep us posted how it goes. Thanks!

**Comment by securisec** [comment]
I am not certain if it is fixed in the latest pulled container. 

I am getting this error when executing a lambda function via api gateway in the way as stated in the OP. 

```
{
  "error": "the JSON object must be str, bytes or bytearray, not Response"
}
```

It works fine in production, and my lambda functions are written in go, being deployed both locally and prod via terraform, and a quick googling shows this is a python error, and the only thing python is localstack. So my triage is that this is a localstack issue.

**Comment by whummer** [comment]
Thanks for the update @securisec . Just to confirm - you are only seeing this behavior in Pro, and not in the community version, is that right? Do you think you could help us with a small, isolated example to reproduce this issue? This would be super helpful, thanks in advance.

**Comment by securisec** [comment]
Thanks @whummer . I will try to put together a small repo that will demonstrate this behavior. But I have been able to narrow down the issue to the authroizer lambda when this error is triggered. 

When running on v 0.12.6, everything works fine (on community) as expected. 

But when I switch to latest and use pro (only time the authroizer function gets executed), i am getting the error.

I set the DEBUG envar

**Comment by securisec** [comment]
### TLDR: It is the authorizer function that is causing the issue. In `pro`, the authroizer function is supposed to trigger and pass correctly, but it is crashing. 

I will work on setting up a small repo for you to test this out @whummer , but in the meantime, this is what I have tried.

Switch back to version 0.12.6 community, and everything works as expected (the api gateway lambda triggers fine) This is also working because the authroizer function does not get triggered in the community version. 

But when I switch to pro (because my stack is using a custom lambda authorizer, I am getting the error. In short, I set the `DEBUG` envar and was able to capture the follow exception message:

```
localstack    |  Traceback (most recent call last):
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_api.py", line 586, in run_lambda
localstack    |     result = LAMBDA_EXECUTOR.execute(func_arn, func_details, event, context=context,
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 176, in execute
localstack    |     return do_execute()
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 168, in do_execute
localstack    |     return _run(func_arn=func_arn)
localstack    |   File "/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py", line 149, in wrapped
localstack    |     raise e
localstack    |   File "/opt/code/localstack/localstack/utils/cloudwatch/cloudwatch_util.py", line 145, in wrapped
localstack    |     result = func(*args, **kwargs)
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 159, in _run
localstack    |     raise e
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 147, in _run
localstack    |     result = self._execute(func_arn, func_details, event, context, version)
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 325, in _execute
localstack    |     result = self.run_lambda_executor(cmd, stdin, env_vars=environment, func_details=func_details)
localstack    |   File "/opt/code/localstack/.venv/lib/python3.8/site-packages/localstack_ext/services/awslambda/lambda_extended.py", line 329, in run_lambda_executor
localstack    |   File "/opt/code/localstack/localstack/services/awslambda/lambda_executors.py", line 231, in run_lambda_executor
localstack    |     raise InvocationException('Lambda process returned error status code: %s. Result: %s. Output:\n%s' %
localstack    | localstack.services.awslambda.lambda_executors.InvocationException: Lambda process returnederror status code: 1. Result: {"errorType":"errorString","errorMessage":"Unauthorized"}. Output:
localstack    | WARNING: Localhost DNS setting (--dns=127.0.0.1) may fail in containers.
localstack    | START RequestId: 4a9f1c06-bed3-16e9-01f6-b6e2d5634526 Version: $LATEST
localstack    | END RequestId: 4a9f1c06-bed3-16e9-01f6-b6e2d5634526
localstack    | REPORT RequestId: 4a9f1c06-bed3-16e9-01f6-b6e2d5634526	Init Duration: 1252.24 ms	Duration: 440.67 ms	Billed Duration: 500 ms	Memory Size: 1536 MB	Max Memory Used: 62 MB
localstack    | INIT: Starting daemons...
localstack    | -----
localstack    | Starting XRay server loop on UDP port 2000
localstack    | Starting DNS server loop on UDP port 53
localstack    |
localstack    |
localstack    | 2021-04-18T18:27:22:INFO:localstack_ext.services.apigateway.authorizers: Received authorizer result: <Response 2387 bytes [500 INTERNAL SERVER ERROR]>
```

follow by the following message on the api response (which makes sense as the authroizer function is failing)
```
{
  "error": "the JSON object must be str, bytes or bytearray, not Response"
}
```

I did not test this on `pro` with the authroizer lambda disabled. But I did test in prod, and the authorizer function is working as intended and is not the cause of the issue.

**Comment by bassteen** [comment]
We are seeying a simular issue having deployed localstack in a k8s cluster using the helm chart. This is the error we are seeying:

`2021-04-22T12:05:38:WARNING:localstack.services.awslambda.lambda_executors: Empty event body specified for invocation of Lambda "arn:aws:lambda:eu-west-1:000000000000:function:my-function"
2021-04-22T12:05:46:INFO:localstack_ext.services.awslambda.lambda_launcher: HOST_TMP_FOLDER not configured - this may break Lambda Docker container startup
2021-04-22T12:05:46:INFO:localstack.services.awslambda.lambda_executors: Running lambda cmd: docker run --user=root --cap-add=NET_ADMIN --entrypoint=/tmp/c8d05a76.sh --dns 127.0.0.1 -e LOCALSTACK_DEBUG=1 -v /tmp/localstack:/tmp -i  -v "/tmp/localstack/zipfile.39c336aa":/var/task -e AWS_REGION="$AWS_REGION" -e DOCKER_LAMBDA_USE_STDIN="$DOCKER_LAMBDA_USE_STDIN" -e LOCALSTACK_HOSTNAME="$LOCALSTACK_HOSTNAME" -e EDGE_PORT="$EDGE_PORT" -e _HANDLER="$_HANDLER" -e AWS_LAMBDA_FUNCTION_TIMEOUT="$AWS_LAMBDA_FUNCTION_TIMEOUT" -e AWS_LAMBDA_FUNCTION_NAME="$AWS_LAMBDA_FUNCTION_NAME" -e AWS_LAMBDA_FUNCTION_VERSION="$AWS_LAMBDA_FUNCTION_VERSION" -e AWS_LAMBDA_FUNCTION_INVOKED_ARN="$AWS_LAMBDA_FUNCTION_INVOKED_ARN" -e AWS_LAMBDA_COGNITO_IDENTITY="$AWS_LAMBDA_COGNITO_IDENTITY" -e NODE_TLS_REJECT_UNAUTHORIZED="$NODE_TLS_REJECT_UNAUTHORIZED"   --rm "localstack/lambda:nodejs12.x" "index.handler"
2021-04-22T12:05:51:DEBUG:localstack.services.awslambda.lambda_executors: Lambda arn:aws:lambda:eu-west-1:000000000000:function:my-function result / log output:

> WARNING: Localhost DNS setting (--dns=127.0.0.1) may fail in containers.
> docker: Error response from daemon: OCI runtime create failed: container_linux.go:367: starting container process caused: exec: "/tmp/c8d05a76.sh": stat /tmp/c8d05a76.sh: no such file or directory: unknown.`

When troubleshooting it seems that files are missing when the volume is mounted. e.g. /tmp/localstack contains:
`drwxr-xr-x    2 localsta localsta      4096 Apr 22 11:07 es_backup
drwxr-xr-x    2 root     root          4096 Apr 22 11:20 __pycache__
-rw-r--r--    1 root     root          5412 Apr 22 12:04 server.test.pem
drwxr-xr-x    3 root     root          4096 Apr 22 12:04 azure-api-specs
-rw-r--r--    1 root     root          1703 Apr 22 12:04 server.test.pem.key
-rw-r--r--    1 root     root          3707 Apr 22 12:04 server.test.pem.crt
drwxr-xr-x    2 root     root          4096 Apr 22 12:05 zipfile.39c336aa
-rwxrwxrwx    1 root     root          9606 Apr 22 12:05 c8d05a76.sh`

but when we mount that volume in the docker container we only see the zipfile
`bash-5.0# docker run --user root -v /tmp/localstack:/tmp -it --entrypoint /bin/bash localstack/lambda:nodejs12.x
bash-4.2# ls -ltr /tmp/
total 0
drwxr-xr-x 2 root root 40 Apr 22 12:05 zipfile.39c336aa`

**Comment by whummer** [comment]
Thanks for reporting @bassteen . Can you please share your configuration details - please note that `HOST_TMP_FOLDER` needs to be configured properly if you're using `LAMBDA_DOCKER_REMOTE=0` (as indicated by the INFO message - should probably be a WARNING message, actually). 

Hope that helps - please keep us posted how it goes, happy to look more into it if you can provide your configuration and a small reproducible test case. Thanks!

**Comment by bassteen** [comment]
Sure. This is our values file which we use to deploy localstack:
[values copy.txt](https://github.com/localstack/localstack/files/6358946/values.copy.txt)

then testing by running:
```
alias awslocal="aws --endpoint-url=http://someurl.localstack.cloud"
awslocal iam create-role --role-name lambda-ex --assume-role-policy-document file://trust-policy.json
awslocal iam attach-role-policy --role-name lambda-ex --policy-arn arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\n
awslocal lambda create-function --function-name my-function --zip-file fileb://function.zip --handler index.handler --runtime nodejs12.x --role arn:aws:iam::000000000000:role/lambda-ex
awslocal lambda invoke --function-name my-function out --log-type Tail
```

where function.zip just contains a hello world lambda function.

**Comment by whummer** [comment]
Thanks for sharing @bassteen . Can you please try to configure `LAMBDA_DOCKER_REMOTE=1` in the pod environment? (This should be the default, but apparently it is not getting applied in this case).

**Comment by nexuschakri** [comment]
I have identical problem like that of securisec. Is there a working docker-compose.yml that solves this issue. The moment I include LOCALSTACK_API_KEY variable to try out a 14 day PRO version in the docker-compose.yml "lambda invoke" fails totally with the following error msg.
{
    "errorMessage": "Lambda process returned error status code: 127. Result: . Output:\nWARNING: Localhost DNS setting (--dns=127.0.0.1) may fail in containers.\ndocker: Error response from daemon: OCI runtime create failed: container_linux.go:367: starting container process caused: exec: \"/tmp/78221821.sh\": stat /tmp/78221821.sh: no such file or directory: unknown.\ntime=\"2021-05-25T07:16:00Z\" level=error msg=\"error getting events from daemon: context canceled\"",
    "errorType": "InvocationException", ... }

**Comment by delitescere** [comment]
@whummer could this be related to the [change of name for managed Lambda policies](https://docs.aws.amazon.com/lambda/latest/dg/security_iam_troubleshoot.html#security_iam_troubleshoot-admin-deprecation) that took effect Mar 1, 2021?

**Comment by whummer** [comment]
Thanks for the update @nexuschakri . Does the Lambda issue still persist with the `latest` Docker image?

Can you please share your `docker-compose.yml` file, share the output of this command:
```
$ pip install --upgrade --no-cache localstack localstack-ext
$ localstack config validate
```
Thanks!

**Comment by whummer** [comment]
Closing this issue due to inactivity. Please let us know if the problem persists @nexuschakri @securisec . Thanks!
